src/storage.rs

```diff
@@ -96,28 +116,54 @@ pub trait Storage {
 /// The Memory Storage Core instance holds the actual state of the storage struct. To access this
 /// value, use the `rl` and `wl` functions on the main MemStorage implementation.
 pub struct MemStorageCore {
-    hard_state: HardState,
-    snapshot: Snapshot,
-    // TODO: maybe vec_deque
+    raft_state: RaftState,
     // entries[i] has raft log position i+snapshot.get_metadata().get_index()
     entries: Vec<Entry>,
+    // Metadata of the last snapshot received.
+    snapshot_metadata: SnapshotMetadata,
 }
 
 impl Default for MemStorageCore {
     fn default() -> MemStorageCore {
         MemStorageCore {
-            // When starting from scratch populate the list with a dummy entry at term zero.
-            entries: vec![Entry::new()],
-            hard_state: HardState::new(),
-            snapshot: Snapshot::new(),
+            raft_state: Default::default(),
+            entries: vec![],
+            // Every time a snapshot is applied to the storage, the metadata will be stored here.
+            snapshot_metadata: Default::default(),
         }
     }
 }
 
 impl MemStorageCore {
     /// Saves the current HardState.
     pub fn set_hardstate(&mut self, hs: HardState) {
-        self.hard_state = hs;
+        self.raft_state.hard_state = hs;
+    }
+
+    /// Get the hard state.
+    pub fn hard_state(&self) -> &HardState {
+        &self.raft_state.hard_state
+    }
+
+    /// Commit to an index.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is no such entry in raft logs.
+    pub fn commit_to(&mut self, index: u64) -> Result<()> {
```

Comment:
```suggestion
    pub fn commit_entry(&mut self, index: u64) -> Result<()> {
```

Unless this is public API elsewhere

Reply:
We have `apply_to` elsewhere so that in this PR let's keep it `commit_to`? We can change all of them in an another PR later.